### PR TITLE
Added json tags to APS, formatted

### PR DESCRIPTION
**Pull request by willfaught**
APS json tags allow unmarshaling APSes from storage.

### CODE REVIEW

File: feedback_test.go [hunk 1]
```diff
@@ -78,7 +79,7 @@ var _ = Describe("Feedback", func() {
 				c := f.Receive()
 
 				r := 0
-				for _ = range c {
+				for range c {
```
Comment (nathany) [inline feedback_test.go]: Note: This change will require Go 1.4+

File: notification.go [hunk 2]
```diff
@@ -53,12 +53,12 @@ func (a *Alert) isZero() bool {
 }
 
 type APS struct {
-	Alert            Alert
-	Badge            *int // 0 to clear notifications, nil to leave as is.
-	Sound            string
-	ContentAvailable int
+	Alert            Alert  `json:"alert,omitempty"`
```
Comment (nathany) [inline notification.go]: These are unnecessary because we're using MarshalJSON.

**Comment by willfaught** [comment]
The description says "unmarshaling". This is for when you're storing the APS you want to use later. The json package will still use MarshalJSON.

Sorry, I didn't know pre-1.4 compatibility was a concern. I just blindly ran gofmt. I can remove the formatting.

**Comment by willfaught** [comment]
Updated to remove the formatting.

**Comment by nathany** [comment]
Actually I'm not sure if pre-1.4 compatibility is a concern or not. Just leaving a note.

I should note that @bdotdub is the project owner, I'm just trying to help out.

**Comment by tylrtrmbl** [comment]
I don't really support using tags to deserialize and a custom method to serialize. I'd rather those be symmetric; either use tags for both, or create an `UnmarshalJSON` method to complement `MarshalJSON`.

If you do tags, we'd have to make `Alert` a pointer as discussed in #28. That's the route I'd take for this pull request.

**Comment by willfaught** [comment]
Agreed, it would be better if it were symmetric.

I didn't want to break the API without knowing the project's compatibility policy. I'd rather not write an UnmarshalJSON. I can amend my PR to change Alert to a pointer type if that's acceptable.

**Comment by tylrtrmbl** [comment]
Ballsy, but I think there's room for that. :smile:

**Comment by bdotdub** [comment]
Another :+1: for symmetric. I think having `Alert` be a pointer makes sense – sounds like we all agree it's more semantically correct.

**Comment by nathany** [comment]
While we're changing the API for Alert, I should point out the [official docs](https://developer.apple.com/library/ios/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Chapters/ApplePushService.html#//apple_ref/doc/uid/TP40008194-CH100-SW1) for the Notification Payload also allow alert to be a string.

>  You may specify a string as the value of alert or a dictionary as its value. If you specify a string, it becomes the message text of an alert.

I'm not sure if that's something we want or need to support, but maybe it's something we should think about.

**Comment by willfaught** [comment]
@nathany anachronistic/apns uses interface{} for Alert, which seemed reasonable to me.

**Comment by nathany** [comment]
ok.

**Comment by tylrtrmbl** [comment]
I'm a fan of #35's approach.

As for serialization for storage, that's pretty highly app-specific. This could go in databases, a binary format, etc. I would put the individual fields into a postgres db. Also, as fields change, this approach to storage is risky. If we rename a field, recovering the struct from storage would fail. This is why storage should be:

- versioned
- app-specific

**Comment by nathany** [comment]
Agree with @tylrtrmbl here. I think you're better off using a JSON struct in your app for serialization, ensuring that it doesn't change when a dependency changes.

**Pull request closed without merging**